src/decoder/lattice-incremental-decoder.h

```diff
@@ -0,0 +1,602 @@
+// limitations under the License.
+
+#ifndef KALDI_DECODER_LATTICE_INCREMENTAL_DECODER_H_
+#define KALDI_DECODER_LATTICE_INCREMENTAL_DECODER_H_
+
+#include "util/stl-utils.h"
+#include "util/hash-list.h"
+#include "fst/fstlib.h"
+#include "itf/decodable-itf.h"
+#include "fstext/fstext-lib.h"
+#include "lat/determinize-lattice-pruned.h"
+#include "lat/kaldi-lattice.h"
+#include "decoder/grammar-fst.h"
+#include "lattice-faster-decoder.h"
+
+namespace kaldi {
+
+struct LatticeIncrementalDecoderConfig {
+  BaseFloat beam;
+  int32 max_active;
+  int32 min_active;
+  BaseFloat lattice_beam;
+  int32 prune_interval;
+  int32 determinize_delay;
+  int32 determinize_max_active;
+  bool redeterminize;
+  bool determinize_lattice; // not inspected by this class... used in
+                            // command-line program.
+  BaseFloat beam_delta;     // has nothing to do with beam_ratio
+  BaseFloat hash_ratio;
+  BaseFloat
+      prune_scale; // Note: we don't make this configurable on the command line,
+                   // it's not a very important parameter.  It affects the
+                   // algorithm that prunes the tokens as we go.
+  // Most of the options inside det_opts are not actually queried by the
+  // LatticeIncrementalDecoder class itself, but by the code that calls it, for
+  // example in the function DecodeUtteranceLatticeIncremental.
+  int32 max_word_id; // for GetLattice
+  fst::DeterminizeLatticePhonePrunedOptions det_opts;
+
+  LatticeIncrementalDecoderConfig()
+      : beam(16.0),
+        max_active(std::numeric_limits<int32>::max()),
+        min_active(200),
+        lattice_beam(10.0),
+        prune_interval(25),
+        determinize_delay(25),
+        determinize_max_active(std::numeric_limits<int32>::max()),
+        redeterminize(false),
+        determinize_lattice(true),
+        beam_delta(0.5),
+        hash_ratio(2.0),
+        prune_scale(0.1),
+        max_word_id(1e7) {}
+  void Register(OptionsItf *opts) {
+    det_opts.Register(opts);
+    opts->Register("beam", &beam, "Decoding beam.  Larger->slower, more accurate.");
+    opts->Register("max-active", &max_active,
+                   "Decoder max active states.  Larger->slower; "
+                   "more accurate");
+    opts->Register("min-active", &min_active, "Decoder minimum #active states.");
+    opts->Register("lattice-beam", &lattice_beam,
+                   "Lattice generation beam.  Larger->slower, "
+                   "and deeper lattices");
+    opts->Register("prune-interval", &prune_interval,
+                   "Interval (in frames) at "
+                   "which to prune tokens");
+    opts->Register("determinize-delay", &determinize_delay,
+                   "delay (in frames) at "
+                   "which to incrementally determinize lattices");
+    opts->Register("determinize-max-active", &determinize_max_active,
+                   "This option is to adaptively decide --determinize-delay. "
```

Comment:
This does not do a very good job of documenting the option.  You don't clarify, for example, whether you are
supposed to set both of them or just one, or if this overrides the other one.  You don't say what a typical value mught be.   Also, for me this option is a bit problematic because it introduces a lot of uncertainty ; you would not be confident that the delay or memory usage would stay reasonably bounded.